Image loader tool

alvaroaleman opened the pull request:
**What this PR does / why we need it**:

Downloads, retags and uploads all docker images we currently deploy via the clusterController

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partially resolves #1045

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

alvaroaleman committed:
Add script+tool to parse images from code

alvaroaleman committed:
Fetch all tags for the images

alvaroaleman committed:
Download images

alvaroaleman committed:
Filter duplicate images

alvaroaleman committed:
Retag and push images

alvaroaleman committed:
Make code less wordy

alvaroaleman committed:
Fix linting

alvaroaleman committed:
Export statefulset and deployment creators

alvaroaleman committed:
Fetch images from creator funcs

alvaroaleman committed:
Use deployment/ss creator funcs

alvaroaleman committed:
Merge branch 'master' into script-download-all-images

alvaroaleman committed:
Create mockapi with required components

alvaroaleman committed:
Move secret cration into func

alvaroaleman committed:
Put mocknamespace name into constant

alvaroaleman committed:
Remove unused code

alvaroaleman committed:
Add printonly mode and execute in ci to avoid code rotting

alvaroaleman committed:
Fix linting errors

alvaroaleman committed:
Populate golang image via template and update to 1.10.2

alvaroaleman committed:
Revert "Populate golang image via template and update to 1.10.2"

This reverts commit f57ea1b26d8566d0f4a28241371fb18fcb1880d6.

alvaroaleman committed:
Test image-loader via go test instead of by executing binary

alvaroaleman committed:
Fix import ordering

alvaroaleman committed:
Remove unneded named return var

mrIncompetent commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+	"github.com/kubermatic/kubermatic/api/pkg/resources"
+
+	"github.com/golang/glog"
+	corev1 "k8s.io/api/core/v1"
```
We normally group imports:
- go-sdk
- others
- kubermatic
- k8s

mrIncompetent commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+	var masterResources string
+	var requestedVersion string
+	var registryName string
+	var printOnly bool
```
Should be moved out of the function and into a `var` block

mrIncompetent commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+	var registryName string
+	var printOnly bool
+
+	flag.StringVar(&masterResources, "master-resources", "../config/kubermatic/static/master/versions.yaml", "")
```
Why is this flag called `master-resources` when the default points to the `versions.yaml` ?

alvaroaleman replied:
The flagname is taken over from the cluster-controller cmd

mrIncompetent replied:
Right, i think by now we could actually replace that flag by dedicated version&update flags.
Though something for another PR

mrIncompetent commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+
+	flag.StringVar(&masterResources, "master-resources", "../config/kubermatic/static/master/versions.yaml", "")
+	flag.StringVar(&requestedVersion, "version", "", "")
+	flag.StringVar(&registryName, "registry-name", "", "Name of the registry to push to")
```
Please add an example which shows the format. Like `docker.io|quay.io`

mrIncompetent commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+	flag.Parse()
+
+	if registryName == "" && !printOnly {
+		glog.Fatalf("Registry name must not be empty!")
```
Rather write `registry-name must not be empty` So there's no confusion about what `Registry` is

mrIncompetent commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+
+	versions, err := version.LoadVersions(masterResources)
+	if err != nil {
+		glog.Fatalf("Error loading versions: %v", err)
```
include the path to the file in the error message. Unless that's already in `err`

mrIncompetent commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+	if requestedVersion == "" {
+		glog.Infof("No version passed, downloading images for all available versions...")
+		for version := range versions {
+			returnedImages, err := getImagesForVersion(versions, version)
```
We should log what versions we're processing

mrIncompetent commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+func pushImages(imageTagList []string) error {
+	for _, image := range imageTagList {
+		glog.Infof("Pushing image %s", image)
+		if out, err := exec.Command("docker", "push", image).CombinedOutput(); err != nil {
```
Log the command you're going to execute

Other thing:
Should we add a retry logic?

alvaroaleman replied:
No, I'd argue the retry logic will be the user executing this tool again

mrIncompetent replied:
sgtm

mrIncompetent commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+		imageSplitted := strings.Split(image, "/")
+		retaggedImageName := fmt.Sprintf("%s/%s", registryName, strings.Join(imageSplitted[1:], "/"))
+		glog.Infof("Tagging image %s as %s", image, retaggedImageName)
+		if out, err := exec.Command("docker", "tag", image, retaggedImageName).CombinedOutput(); err != nil {
```
Log the command you're going to execute

mrIncompetent commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+func downloadImages(images []string) error {
+	for _, image := range images {
+		glog.Infof("Downloading image '%s'...\n", image)
+		if out, err := exec.Command("docker", "pull", image).CombinedOutput(); err != nil {
```
Log the command you're going to execute

Other thing:
Should we add a retry logic?

metalmatze replied:
If you want to do retry logic I can highly recommend:
https://github.com/cenkalti/backoff

mrIncompetent commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+	go serviceInformer.Informer().Run(wait.NeverStop)
+	cache.WaitForCacheSync(wait.NeverStop, configMapInformer.Informer().HasSynced)
+	cache.WaitForCacheSync(wait.NeverStop, secretInformer.Informer().HasSynced)
+	cache.WaitForCacheSync(wait.NeverStop, serviceInformer.Informer().HasSynced)
```
You can replace the above 6 lines with the kubeInformerFactory functions.
```go
kubeInformerFactory.Start()
kubeInformerFactory.WaitForCacheSync()
```

mrIncompetent commented:
Do you want to add a Readme here or are you going to open a separate PR for it?

alvaroaleman commented:
Regarding the readme I thought this would be a good candidate for our new hugo doc system, allhought it obviously makes sense to version it together with the code - We could probably keep it in both places for the time being and at some future point write some exporting logic that exports docs that are versioned here into the docs repo?

mrIncompetent commented:
Regarding the docs:
sgtm

@cbrgm where do we have to write our documentation so it gets published via Hugo?
Also where is the documentation available?

cbrgm commented:
Documentation is available here https://github.com/kubermatic/docs/tree/kubermatic-docs.
You simply have to drop markdown files with some annotated meta data on top of the file into the `/content` folder. After that you can run the `hugo` command and it will generate all static website files into a folder called `/public`.

p0lyn0mial commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+	cache.WaitForCacheSync(wait.NeverStop, secretInformer.Informer().HasSynced)
+	cache.WaitForCacheSync(wait.NeverStop, serviceInformer.Informer().HasSynced)
+
+	return &resources.TemplateData{Version: version,
```
how do we make use of `Version` ?
I think that `resources.TemplateData` doesn't have `Version` member.

mrIncompetent replied:
Not anymore. The controller exclusively takes the version from the Cluster object.
The version/update handling got some polishing in https://github.com/kubermatic/kubermatic/pull/1095

p0lyn0mial commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+
+func retagImages(registryName string, imageTagList []string) (retaggedImages []string, err error) {
+	for _, image := range imageTagList {
+		imageSplitted := strings.Split(image, "/")
```
why not to check the length of `imageSplitted` ?

alvaroaleman replied:
Good point, added it

p0lyn0mial commented on api/cmd/image-loader/main.go:
```diff
@@ -0,0 +1,264 @@
+		"kubelet-client-certificates",
+		"service-account-key"})
+	objects := []runtime.Object{configMapList, secretList, serviceList}
+	client := kubefake.NewSimpleClientset(objects...)
```
nice :)

alvaroaleman committed:
Merge branch 'master' into script-download-all-images

alvaroaleman committed:
Fix import ordering

alvaroaleman committed:
Move vars into dedicated var block

alvaroaleman committed:
Finish merging

alvaroaleman committed:
Name versionsFile parameter versions-file

alvaroaleman committed:
Improve error messages

alvaroaleman committed:
Add logging when collecting images

alvaroaleman committed:
Log executed commands

alvaroaleman committed:
Use kubeInformerFactory to start Informers and wait for their sync

alvaroaleman committed:
Verify imageName contains a registry

alvaroaleman committed:
Add README

alvaroaleman committed:
Fix golint error

alvaroaleman commented:
@mrIncompetent @p0lyn0mial I think I addressed all comments, PTAL. Had to update some of the version loading logic because it got significantly changed in #1095

alvaroaleman committed:
Merge branch 'master' into script-download-all-images
